src/QsCompiler/Compiler/Templates/QirDriverCpp.tt

```diff
@@ -1,16 +1,342 @@
 ﻿<#@ template language="C#" #>
+<#@ import namespace="Microsoft.Quantum.QsCompiler.BondSchemas.EntryPoint" #>
 //------------------------------------------------------------------------------
 // This code was generated by a tool.
 // <auto-generated />
 //------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
 
-// N.B. This is currently just a base non-functional example of a template used
-//      to generate C++ code that works as an entry-point driver for a QIR-based
-//      executable.
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
 
-extern "C" void <#= entryPointOperation.Name #>(); // NOLINT
+#include "CLI11.hpp"
+
+#include "QirContext.hpp"
+#include "QirRuntime.hpp"
+#include "SimFactory.hpp"
+
+using namespace Microsoft::Quantum;
+using namespace std;
+
+struct InteropArray
+{
+    int64_t Size;
+    void* Data;
+
+    InteropArray(int64_t size, void* data) :
+        Size(size),
+        Data(data){}
+};
+
+using RangeTuple = tuple<int64_t, int64_t, int64_t>;
+struct InteropRange
+{
+    int64_t Start;
+    int64_t Step;
+    int64_t End;
+
+    InteropRange() :
+        Start(0),
+        Step(0),
+        End(0){}
+
+    InteropRange(RangeTuple rangeTuple) :
+        Start(get<0>(rangeTuple)),
+        Step(get<1>(rangeTuple)),
+        End(get<2>(rangeTuple)){}
+};
+
+// This is the function corresponding to the QIR entry-point.
+extern "C" void <#= entryPointOperation.Name #>(
+    <# foreach (var arg in entryPointOperation.Arguments) { #>
+        <# if (arg.Type == DataType.IntegerType) { #>
```

Comment:
It would be great to encapsulate some of the logic outside the template itself and keep the code in the template minimal.

One way this can be done is to extend the classes defined in the Bond schema since they are generated as partial classes. For example, an _EntryPointCpp.cs_ file can be created that contain logic specific for C++ generation. More specifically, something like this could be done in that file:
```csharp
    public partial class Argument
    {
        public string CppType() =>
            this.Type switch
            {
                DataType.BoolType => "char",
                DataType.IntType => "int64_t",
                // Add all the rest of the types here.
            };
    }
```
This can be used in the template as `arg.CppType()` which reduces the amount of logic needed in the template itself.

Reply:
done